plutus-ledger-api/src/PlutusLedgerApi/V1/Value.hs

```diff
@@ -368,6 +364,118 @@ split (Value mp) = (negate (Value neg), Value pos) where
+
+For these reasons we decided not to pick the winner (@valueEqualsValue4@) from
+    https://github.com/input-output-hk/plutus/issues/5135#issuecomment-1459327829
+despite the fact that in some cases it performs much better than our solution (while performing
+substantially worse in others).
+
+If the result is 'MatchSuccess', then the two lists are equal.
+If the result is 'MatchFailure', then the two lists are not equal (pointwise at least).
+If the result is @MatchPartial vvs kvs1' kvs2'@, then the two lists have the same keys in the
+  beginning (@length vvs@ of them), but diverge at some point. @vvs@ contains values associated with
+  the matching keys (the first component of each tuple comes from the first list, the second
+  component of each tuple comes from the second list). @kvs1'@ and @kvs2'@ are the first and the
+  second list with their prefixes (those that have matching keys) stripped.
+-}
+matchKVs
+    :: forall k v. Ord k
+    => (v -> Bool) -> (v -> v -> Bool) -> [(k, v)] -> [(k, v)] -> MatchResult k v
+matchKVs ~is0 ~structEqV = go where
+    go :: [(k, v)] -> [(k, v)] -> MatchResult k v
+    -- Spines match, hence it's a 'MatchSuccess' so far.
+    go [] [] = MatchSuccess
+    -- One spine is longer than the other one, but this still can be a 'MatchSuccess' if the
+    -- non-empty lists only consists of empty values.
+    go [] kvs2 = if all (is0 . snd) kvs2 then MatchSuccess else MatchFailure
+    -- Symmetric to the previous case.
+    go kvs1 [] = if all (is0 . snd) kvs1 then MatchSuccess else MatchFailure
+    -- Both spines are non-empty.
+    go kvs1@((k1, v1) : kvs1') kvs2@((k2, v2) : kvs2')
+        -- If keys are equal
+        | k1 == k2 =
+            -- then continue checking equality of spines.
+            case go kvs1' kvs2' of
+                -- If spines are equal, then we can proceed to checking equality of values as by
+                -- this point we've ensured that the keys in the lists match exactly.
+                MatchSuccess -> if structEqV v1 v2 then MatchSuccess else MatchFailure
+                -- If there was a key mismatch down the line, then we cons the values associated
+                -- with @k1@ and @k2@ onto @vvs@ to check their equality after we ensure that keys
+                -- from @kvs1''@ are a permutation of those from @kvs2''@.
+                MatchPartial vvs kvs1'' kvs2'' -> MatchPartial ((v1, v2) : vvs) kvs1'' kvs2''
+                -- A failure stays a failure.
+                MatchFailure -> MatchFailure
+        -- If the keys aren't equal, then maybe the first value is empty, in which case we throw it
+        -- out and proceed.
+        | is0 v1 = go kvs1' kvs2
+        -- Or if the second one is empty, then we throw that one and proceed.
+        | is0 v2 = go kvs1 kvs2'
+        -- Otherwise the keys in the lists have diverged and we return the remaining parts.
+        | otherwise = MatchPartial [] kvs1 kvs2
+
+{-# INLINEABLE eq #-}
```

Comment:
I think this doesn't actually work: there's a problem with non-trailing entries with zero values.  In the code below, `v1 == v2` is `False` but `v2 == v3` is `True`.

```
{-# LANGUAGE OverloadedStrings #-}

import PlutusLedgerApi.V1.Value
import PlutusTx.AssocMap as Map
    
ada :: CurrencySymbol
ada = ""

a :: TokenName
a = "aa"

b :: TokenName
b = "bb"

c :: TokenName
c = "cc"
    
v1 :: Value
v1 = Value $ Map.fromList [(ada, Map.fromList [(a,0),(b,100)])]

v2 :: Value
v2 = Value $ Map.fromList [(ada, Map.fromList [(b,100)])]

v3 :: Value
v3 = Value $ Map.fromList [(ada, Map.fromList [(b,100),(c,0)])]
```

Reply:
Yup, thanks a lot for debugging the issue and providing the example. It's a bug indeed: I forgot to replicate the `is0` logic of `matchKVs` in `eqKVs`. My imaginary tests would catch the bug, but I'm rushing (for PI planning reasons) and imaginary tests don't get run in the real world, unfortunately. I've fixed the bug now and I'll add proper tests in a subsequent PR.